README.md
# YoutubeSourceParserKit 

==================
[![Build Status](https://img.shields.io/travis/movielala/YoutubeSourceParserKit/master.svg)](https://travis-ci.org/movielala/YoutubeSourceParserKit)
![License](https://img.shields.io/badge/license-MIT-blue.svg)
[![Dependencies](https://img.shields.io/badge/dependencies-none-brightgreen.svg)](https://github.com/mobileplayer/mobileplayer-ios)
[![Ready](https://badge.waffle.io/movielala/YoutubeSourceParserKit.png?label=Ready&title=Ready)](https://waffle.io/movielala/YoutubeSourceParserKit)
[![StackOverflow](https://img.shields.io/badge/StackOverflow-Ask%20a%20question!-blue.svg)](http://stackoverflow.com/questions/ask?tags=YoutubeSourceParserKit+ios+swift)
[![Cocoapods](https://img.shields.io/cocoapods/v/YoutubeSourceParserKit.svg)](https://img.shields.io/cocoapods/v/YoutubeSourceParserKit.svg)
[![Join the chat at https://gitter.im/mobileplayer/mobileplayer-ios](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/movielala/YoutubeSourceParserKit)


YouTube Video Link Parser for Swift

##Introduction

__Requires iOS 8 or later and Xcode 7.0+__<br/>
Swift support uses dynamic frameworks and is therefore only supported on iOS > 8.

##Installation

To install via CocoaPods add this line to your `Podfile`.
```
use_frameworks!
```
and
```
pod 'YoutubeSourceParserKit'
```

Then, run the following command:

```$ pod install```

##Usage

```swift
import YoutubeSourceParserKit
```

```swift
let testURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")!
    Youtube.h264videosWithYoutubeURL(testURL) { (videoInfo, error) -> Void in
      if let videoURLString = videoInfo?["url"] as? String,
        videoTitle = videoInfo?["title"] as? String {
          print("\(videoTitle)")
          print("\(videoURLString)")
      }
    }
```

```
videoInfo output:
```
```json
{
    "title": "[Video Title]",
    "isStream": 0,
    "quality": "hd720",
    "itag": 22,
    "fallback_host": "tc.v20.cache2.googlevideo.com",
    "url": "http://[Source URL]"
}
```

##MPMoviePlayerController Usage

![alt tag](http://s10.postimg.org/5j1mristl/i_OS_Simulator_Screen_Shot_Jul_12_2015_14_33_02.png)

```swift
import UIKit
import YoutubeSourceParserKit
import MediaPlayer

class ViewController: UIViewController {

  let moviePlayer = MPMoviePlayerController()

  override func viewDidLoad() {
    super.viewDidLoad()
    moviePlayer.view.frame = view.frame
    view.addSubview(moviePlayer.view)
    moviePlayer.fullscreen = true
    let youtubeURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")!
    playVideoWithYoutubeURL(youtubeURL)
  }

  func playVideoWithYoutubeURL(url: NSURL) {
    Youtube.h264videosWithYoutubeURL(url, completion: { (videoInfo, error) -> Void in
      if let
        videoURLString = videoInfo?["url"] as? String,
        videoTitle = videoInfo?["title"] as? String {
          self.moviePlayer.contentURL = NSURL(string: videoURLString)
      }
    })
  }
}
```


CHANGELOG.md
#### Version 0.2.4

* Rename 'YoutubeSourceParserKit'
* Add tvOS support

#### Version 0.2.3

* Supporting Swift 2.0

#### Version 0.2.2

* Add Changelog file

#### Version 0.2.1

* Fixed gcd bug

#### Version 0.2.0

* Add gcd to parser operations

#### Version 0.1.9

* Updated project setting and podspec file

#### Version 0.1.8

* Supporting for iOS7

#### Version 0.1.7

* Add test conditions and update podspec file

#### Version 0.1.6

* Update README.md

#### Version 0.1.5

* Update podspec file

#### Version 0.1.4

* Update podspec file

#### Version 0.1.3

* Add MPMoviePlayerController Usage

#### Version 0.1.2

* Update podspec file

#### Version 0.1.1

* Add youtube-parser class

#### Version 0.1.0

* Initial version


AppDelegate.swift
//
//  AppDelegate.swift
//  youtube-parser
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
  
  var window: UIWindow?
  func application(application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
    let testURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")!
    Youtube.h264videosWithYoutubeURL(testURL) { (videoInfo, error) -> Void in
      if let videoURLString = videoInfo?["url"] as? String,
        videoTitle = videoInfo?["title"] as? String {
          print("\(videoTitle)")
          print("\(videoURLString)")
      }
    }
    return true
  }
  
  func applicationWillResignActive(application: UIApplication) { }
  
  func applicationDidEnterBackground(application: UIApplication) { }
  
  func applicationWillEnterForeground(application: UIApplication) { }
  
  func applicationDidBecomeActive(application: UIApplication) { }
  
  func applicationWillTerminate(application: UIApplication) { }
}


YoutubeSourceParserKit/Youtube.swift
//
//  Youtube.swift
//  youtube-parser
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit

public extension NSURL {
  /**
  Parses a query string of an NSURL

  @return key value dictionary with each parameter as an array
  */
  func dictionaryForQueryString() -> [String: AnyObject]? {
    if let query = self.query {
      return query.dictionaryFromQueryStringComponents()
    }

    // Note: find youtube ID in m.youtube.com "https://m.youtube.com/#/watch?v=1hZ98an9wjo"
    let result = absoluteString.componentsSeparatedByString("?")
    if result.count > 1 {
      return result.last?.dictionaryFromQueryStringComponents()
    }
    return nil
  }
}

public extension NSString {
  /**
  Convenient method for decoding a html encoded string
  */
  func stringByDecodingURLFormat() -> String {
    let result = self.stringByReplacingOccurrencesOfString("+", withString:" ")
    return result.stringByRemovingPercentEncoding!
  }

  /**
  Parses a query string

  @return key value dictionary with each parameter as an array
  */
  func dictionaryFromQueryStringComponents() -> [String: AnyObject] {
    var parameters = [String: AnyObject]()
    for keyValue in componentsSeparatedByString("&") {
      let keyValueArray = keyValue.componentsSeparatedByString("=")
      if keyValueArray.count < 2 {
        continue
      }
      let key = keyValueArray[0].stringByDecodingURLFormat()
      let value = keyValueArray[1].stringByDecodingURLFormat()
      parameters[key] = value
    }
    return parameters
  }
}

public class Youtube: NSObject {
  static let infoURL = "http://www.youtube.com/get_video_info?video_id="
  static var userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4"
  /**
  Method for retrieving the youtube ID from a youtube URL

  @param youtubeURL the the complete youtube video url, either youtu.be or youtube.com
  @return string with desired youtube id
  */
  public static func youtubeIDFromYoutubeURL(youtubeURL: NSURL) -> String? {
    if let
      youtubeHost = youtubeURL.host,
      youtubePathComponents = youtubeURL.pathComponents {
        let youtubeAbsoluteString = youtubeURL.absoluteString
        if youtubeHost == "youtu.be" as String? {
          return youtubePathComponents[1]
        } else if youtubeAbsoluteString.rangeOfString("www.youtube.com/embed") != nil {
          return youtubePathComponents[2]
        } else if youtubeHost == "youtube.googleapis.com" ||
          youtubeURL.pathComponents!.first == "www.youtube.com" as String? {
            return youtubePathComponents[2]
        } else if let
          queryString = youtubeURL.dictionaryForQueryString(),
          searchParam = queryString["v"] as? String {
            return searchParam
        }
    }
    return nil
  }
  /**
  Method for retreiving a iOS supported video link

  @param youtubeURL the the complete youtube video url
  @return dictionary with the available formats for the selected video
  
  */
  public static func h264videosWithYoutubeID(youtubeID: String) -> [String: AnyObject]? {
    let urlString = String(format: "%@%@", infoURL, youtubeID) as String
    let url = NSURL(string: urlString)!
    let request = NSMutableURLRequest(URL: url)
    request.timeoutInterval = 5.0
    request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
    request.HTTPMethod = "GET"
    var responseString = NSString()
    let session = NSURLSession(configuration: NSURLSessionConfiguration.defaultSessionConfiguration())
    let group = dispatch_group_create()
    dispatch_group_enter(group)
    session.dataTaskWithRequest(request, completionHandler: { (data, response, _) -> Void in
      if let data = data as NSData? {
        responseString = NSString(data: data, encoding: NSUTF8StringEncoding)!
      }
      dispatch_group_leave(group)
    }).resume()
    dispatch_group_wait(group, DISPATCH_TIME_FOREVER)
    let parts = responseString.dictionaryFromQueryStringComponents()
    if parts.count > 0 {
      var videoTitle: String = ""
      var streamImage: String = ""
      if let title = parts["title"] as? String {
        videoTitle = title
      }
      if let image = parts["iurl"] as? String {
        streamImage = image
      }
      if let fmtStreamMap = parts["url_encoded_fmt_stream_map"] as? String {
        // Live Stream
        if let _: AnyObject = parts["live_playback"]{
          if let hlsvp = parts["hlsvp"] as? String {
            return [
              "url": "\(hlsvp)",
              "title": "\(videoTitle)",
              "image": "\(streamImage)",
              "isStream": true
            ]
          }
        } else {
          let fmtStreamMapArray = fmtStreamMap.componentsSeparatedByString(",")
          for videoEncodedString in fmtStreamMapArray {
            var videoComponents = videoEncodedString.dictionaryFromQueryStringComponents()
            videoComponents["title"] = videoTitle
            videoComponents["isStream"] = false
            return videoComponents as [String: AnyObject]
          }
        }
      }
    }
    return nil
  }
  
  /**
  Block based method for retreiving a iOS supported video link

  @param youtubeURL the the complete youtube video url
  @param completeBlock the block which is called on completion

  */
  public static func h264videosWithYoutubeURL(youtubeURL: NSURL,completion: ((
    videoInfo: [String: AnyObject]?, error: NSError?) -> Void)?) {
      let priority = DISPATCH_QUEUE_PRIORITY_BACKGROUND
      dispatch_async(dispatch_get_global_queue(priority, 0)) {
        if let youtubeID = self.youtubeIDFromYoutubeURL(youtubeURL), videoInformation = self.h264videosWithYoutubeID(youtubeID) {
          dispatch_async(dispatch_get_main_queue()) {
            completion?(videoInfo: videoInformation, error: nil)
          }
        }else{
          dispatch_async(dispatch_get_main_queue()) {
            completion?(videoInfo: nil, error: NSError(domain: "com.player.youtube.backgroundqueue", code: 1001, userInfo: ["error": "Invalid YouTube URL"]))
          }
        }
      }
    }
  }


youtube-parserTests/youtube_parserTests.swift
//
//  youtube_parserTests.swift
//  youtube-parserTests
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit
import XCTest

class youtube_parserTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }

  func testStringByDecodingURLFormat() {
    let originalString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ+XUFpdwbfqqQ"
    let testString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ XUFpdwbfqqQ"
    let comparsionString = originalString.stringByDecodingURLFormat()
    XCTAssertNotNil(comparsionString, "comparsionString is nil")
    XCTAssertEqual(comparsionString, testString, "String decoding failed")
  }

  func testDictionaryFromQueryStringComponents() {
    let sampleLink = "https://www.youtube.com/watch?v=o0jJiB2Ygpg&list=RDo0jJiB2Ygpg"
    let dictionary = sampleLink.dictionaryFromQueryStringComponents() as [String:AnyObject]
    XCTAssertNotNil(dictionary["list"], "url dictionary parse error")
    if let list = dictionary["list"] as? String {
      XCTAssertEqual(list, "RDo0jJiB2Ygpg", "list not equal value")
    }
    if let link = dictionary["https://www.youtube.com/watch?v="] as? String {
      XCTAssertEqual(link, "o0jJiB2Ygpg", "link not equal value")
    }
  }

  func testInvalidURLs1() {
    let sampleLink = NSURL(string: "?v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs2() {
    let sampleLink = NSURL(string: "?v=")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "", "Youtube ID not same")
    }
  }

  func testInvalidURLs3() {
    let sampleLink = NSURL(string: "v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "v=1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs4() {
    let sampleLink = NSURL(string: "v1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "v1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromYoutubeURL() {
    let sampleLink = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromMobileYoutubeURL() {
    let sampleLink = NSURL(string: "https://m.youtube.com/#/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testH264videosWithYoutubeURL() {
    if let videoComponents = Youtube.h264videosWithYoutubeID("1hZ98an9wjo") {
      XCTAssertNotNil(videoComponents, "video component is nil")
      if let itag = videoComponents["itag"] as? String {
        XCTAssertEqual(itag, "22", "itag not equal")
      }
      if let quality = videoComponents["quality"] as? String {
        XCTAssertEqual(quality, "hd720", "quality not equal")
      }
      if let fallback_host = videoComponents["fallback_host"] as? String {
        XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
      }
      if let type = videoComponents["type"] as? String {
        XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
      }
    }
  }

  func testVideoQuality(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["quality"] as? String {
            XCTAssertEqual(quality, "hd720", "quality not equal")
          }
        }
      })
    }
  }

  func testItagControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let itag = info["itag"] as? String {
            XCTAssertEqual(itag, "22", "itag not equal")
          }
        }
      })
    }
  }

  func testVideoTypeControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let type = info["type"] as? String {
            XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
          }
        }
      })
    }
  }

  func testFallbackHostControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let fallback_host = info["fallback_host"] as? String {
            XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
          }
        }
      })
    }
  }

  func testIsLiveVideoTest() {
    if let liveVideoURL = NSURL(string: "https://www.youtube.com/watch?v=rxGoGg7n77A"){
      Youtube.h264videosWithYoutubeURL(liveVideoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          XCTAssertNotNil(info["url"], "live stream url is nil")
        }
      })
    }
  }

  func testInfoUrl(){
    XCTAssertEqual(Youtube.infoURL, "http://www.youtube.com/get_video_info?video_id=", "Info url not correct!")
  }

  func testUserAgent() {
     XCTAssertEqual(Youtube.userAgent, "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4", "User Agent not correct!")
  }

  func testH264videosWithYoutubeURLBlock(){
    let expectation: XCTestExpectation = self.expectationWithDescription("Handler called")
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        expectation.fulfill()
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["itag"] as? String {
            XCTAssertEqual(quality, "22", "itag not equal")
          }
        }
      })
      self.waitForExpectationsWithTimeout(2, handler: nil)
    }

  }
  
}


YoutubeSourceParserKitTests/YoutubeSourceParserKitTests.swift
//
//  youtube_parserTests.swift
//  youtube-parserTests
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit
import XCTest

class YoutubeSourceParserKitTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }

  func testStringByDecodingURLFormat() {
    let originalString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ+XUFpdwbfqqQ"
    let testString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ XUFpdwbfqqQ"
    let comparsionString = originalString.stringByDecodingURLFormat()
    XCTAssertNotNil(comparsionString, "comparsionString is nil")
    XCTAssertEqual(comparsionString, testString, "String decoding failed")
  }

  func testDictionaryFromQueryStringComponents() {
    let sampleLink = "https://www.youtube.com/watch?v=o0jJiB2Ygpg&list=RDo0jJiB2Ygpg"
    let dictionary = sampleLink.dictionaryFromQueryStringComponents() as [String:AnyObject]
    XCTAssertNotNil(dictionary["list"], "url dictionary parse error")
    if let list = dictionary["list"] as? String {
      XCTAssertEqual(list, "RDo0jJiB2Ygpg", "list not equal value")
    }
    if let link = dictionary["https://www.youtube.com/watch?v="] as? String {
      XCTAssertEqual(link, "o0jJiB2Ygpg", "link not equal value")
    }
  }

  func testInvalidURLs1() {
    let sampleLink = NSURL(string: "?v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs2() {
    let sampleLink = NSURL(string: "?v=")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "", "Youtube ID not same")
    }
  }

  func testInvalidURLs3() {
    let sampleLink = NSURL(string: "v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "v=1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs4() {
    let sampleLink = NSURL(string: "v1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "v1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromYoutubeURL() {
    let sampleLink = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromMobileYoutubeURL() {
    let sampleLink = NSURL(string: "https://m.youtube.com/#/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testH264videosWithYoutubeURL() {
    if let videoComponents = Youtube.h264videosWithYoutubeID("1hZ98an9wjo") {
      XCTAssertNotNil(videoComponents, "video component is nil")
      if let itag = videoComponents["itag"] as? String {
        XCTAssertEqual(itag, "22", "itag not equal")
      }
      if let quality = videoComponents["quality"] as? String {
        XCTAssertEqual(quality, "hd720", "quality not equal")
      }
      if let fallback_host = videoComponents["fallback_host"] as? String {
        XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
      }
      if let type = videoComponents["type"] as? String {
        XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
      }
    }
  }

  func testVideoQuality(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["quality"] as? String {
            XCTAssertEqual(quality, "hd720", "quality not equal")
          }
        }
      })
    }
  }

  func testItagControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let itag = info["itag"] as? String {
            XCTAssertEqual(itag, "22", "itag not equal")
          }
        }
      })
    }
  }

  func testVideoTypeControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let type = info["type"] as? String {
            XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
          }
        }
      })
    }
  }

  func testFallbackHostControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let fallback_host = info["fallback_host"] as? String {
            XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
          }
        }
      })
    }
  }

  func testIsLiveVideoTest() {
    if let liveVideoURL = NSURL(string: "https://www.youtube.com/watch?v=rxGoGg7n77A"){
      Youtube.h264videosWithYoutubeURL(liveVideoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          XCTAssertNotNil(info["url"], "live stream url is nil")
        }
      })
    }
  }

  func testInfoUrl(){
    XCTAssertEqual(Youtube.infoURL, "http://www.youtube.com/get_video_info?video_id=", "Info url not correct!")
  }

  func testUserAgent() {
     XCTAssertEqual(Youtube.userAgent, "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4", "User Agent not correct!")
  }

  func testH264videosWithYoutubeURLBlock(){
    let expectation: XCTestExpectation = self.expectationWithDescription("Handler called")
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
      Youtube.h264videosWithYoutubeURL(videoURL, completion: { (videoInfo, error) -> Void in
        expectation.fulfill()
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["itag"] as? String {
            XCTAssertEqual(quality, "22", "itag not equal")
          }
        }
      })
      self.waitForExpectationsWithTimeout(2, handler: nil)
    }

  }
  
}
